WebCmdlets get encoding from BOM

CarloToso opened the pull request:
<!-- Anything that looks like this is a comment and can't be seen after the Pull Request is created. -->

# PR Summary

- Add EncodingHelper to detect the response encoding from the BOM
- Add tests

<!-- Summarize your PR between here and the checklist. -->

## PR Context
using: https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/IO/StreamReader.cs 

fixes #11547

<!-- Provide a little reasoning as to why this Pull Request helps and why you have opened it. -->

## PR Checklist

- [ ] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [ ] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [ ] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed: <!-- Number/link of that issue here -->

CarloToso committed:
Add EncodingHelper

CarloToso committed:
remove const

CarloToso committed:
add stream.length chack

CarloToso committed:
use when

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/StreamHelper.cs:
```diff
@@ -383,9 +383,24 @@ private static string StreamToString(Stream stream, Encoding encoding)
+                stream.ReadExactly(buffer, 0, 4);
+            }
+
+            EncodingHelper.TryDetectEncodingFromBom(buffer, out encoding, out int preambleLength);
```
I would be surprised if this is not already done in .Net. It might be worth looking at HttpClient and other related code.

CarloToso replied:
I read the code in .Net and it doesn't match perfectly with our needs (it doesn't consider UTF32-BE, it checks for the BOM after checking the charset)

iSazonov replied:
> it checks for the BOM after checking the charset

Maybe it is more right behavior? Could you please point the code?

CarloToso replied:
Gets encoding from charset:
https://github.com/dotnet/runtime/blob/bd6cbe3642f51d70839912a6a666e5de747ad581/src/libraries/System.Net.Http/src/System/Net/Http/HttpContent.cs#L194-L219

Gets encoding from bom:
https://github.com/dotnet/runtime/blob/bd6cbe3642f51d70839912a6a666e5de747ad581/src/libraries/System.Net.Http/src/System/Net/Http/HttpContent.cs#L221-L234

iSazonov replied:
The code is 8 years old. So we should follow the logic too.

CarloToso replied:
I followed what was discussed here: https://github.com/PowerShell/PowerShell/issues/11547#issuecomment-636522169

iSazonov replied:
Thanks for pointing the history. It is correct.

Can we use StreamReader with auto encoding detection?
https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/IO/StreamReader.cs,109

CarloToso replied:
@iSazonov I tried using StreamReader with auto encoding detection, what do you think?

```cs
        internal static string DecodeStream(Stream stream, string characterSet, out Encoding encoding)
        {
            bool isDefaultEncoding = false;

            StreamReader reader = new(stream, detectEncodingFromByteOrderMarks: true);

            // reader.CurrentEncoding defaults to UTF8
            encoding = reader.CurrentEncoding;

            if (encoding == Encoding.UTF8)
            {
                isDefaultEncoding = !TryGetEncodingFromCharset(characterSet, out encoding);
            }

            if (isDefaultEncoding)
            {
                reader = new(stream, encoding);

                // We only look within the first 1k characters as the meta element and
                // the xml declaration are at the start of the document
                int bufferLength = (int)Math.Min(reader.BaseStream.Length, 1024);

                char[] buffer = new char[bufferLength];
                reader.ReadBlock(buffer, 0, bufferLength);

                string substring = new(buffer);

                // Check for a charset attribute on the meta element to override the default
                Match match = s_metaRegex.Match(substring);
                
                // Check for a encoding attribute on the xml declaration to override the default
                if (!match.Success)
                {
                    match = s_xmlRegex.Match(substring);
                }
                
                if (match.Success)
                {
                    characterSet = match.Groups["charset"].Value;

                    if (TryGetEncodingFromCharset(characterSet, out Encoding localEncoding))
                    {
                        encoding = localEncoding;
                    }
                }
            }

            return new StreamReader(stream, encoding).ReadToEnd();
        }
```

iSazonov replied:
I guess we can use https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/IO/StreamReader.cs,119
and pass encoding based on characterSet. This will aromatically use BOM if present.

CarloToso committed:
use StreamReader detectEncodingFromByteOrderMarks

CarloToso committed:
fix return

CarloToso committed:
stream leaveOpen: true

CarloToso committed:
add tests

CarloToso committed:
fix codefactor

CarloToso committed:
fix tests

iSazonov commented on test/tools/WebListener/Controllers/EncodingController.cs:
```diff
@@ -51,6 +51,54 @@ public async void Cp936()
             };
             await Response.Body.WriteAsync(body, 0, body.Length);
         }
+        public async void Utf8BOM()
```
```suggestion

        public async void Utf8BOM()
```

CarloToso committed:
reader.Peek()

CarloToso commented:
@iSazonov any insight on the many test failures after the last commit?

iSazonov commented:
@CarloToso I guess changes in EncodingController.cs cause CI fails.

CarloToso committed:
comment out reader.Peek()

CarloToso commented:
@iSazonov it seems the errors were caused by `reader.Peek()`

iSazonov commented:
> @iSazonov it seems the errors were caused by `reader.Peek()`

It only force encoding detection. This says either code, or tests, or both are broken.

CarloToso committed:
try fix

mklement0 commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/StreamHelper.cs:
```diff
@@ -329,83 +329,25 @@ internal static void SaveStreamToFile(Stream stream, string filePath, PSCmdlet c
-
-            stream.Seek(preambleLength, SeekOrigin.Begin);
-            string content = StreamToString(stream, encoding);
+            encoding = reader.CurrentEncoding;
```
@iSazonov, I'm afraid I don't understand the question (I'm not familiar with the code).

iSazonov replied:
@mklement0 I ask about p.3 from https://github.com/PowerShell/PowerShell/issues/11547#issuecomment-636522169
> otherwise, for XML and HTML, respect the encoding specified in the XML declaration

iSazonov replied:
@mklement0 Friendly ping.

mklement0 replied:
@iSazonov, sorry, I forgot to respond: I'm still unclear on the details, but what I suggested in https://github.com/PowerShell/PowerShell/issues/11547#issuecomment-636522169 meant looking no further once a BOM is found. Only if there's none, look at `charset`. Only if there's none, look at the XML declaration / HTML `<meta>` element (the latter seems to be absent from the snippet above - is HTML handled elsewhere?). 
This hinges on explicitly knowing if a BOM is present.

iSazonov replied:
@mklement0 Thanks! What I don't understand is how it should work if the encoding is specified inside the XML/HTML. Now we always perform stream conversion to UTF16 string first, which means subsequent conversion based on attribute from XML will not be correct, right? That means we either have to ignore that XML attribute or change our decoding completely and do it on the original byte stream.

mklement0 replied:
(As for also looking for a `<meta>` element: I didn't read carefully the first time, and I now see that it is also being detected).

Yes, there is a problem with the detection code: it only works properly if the actual encoding is a _single-byte_ encoding that is a _superset_ of ASCII, so it only works with the various OEM and ANSI single-byte Windows encodings and UTF-8, for instance, but _not_ with other Unicode encodings (UTF-16LE, UTF-16BE, ...)

While limiting the detection to ASCII supersets strikes me as sufficient in practice (I don't know what the RFCs say),
additional work is required to detect the other Unicode encodings, which, however would be as simple as doing `substring.Replace("\0", "")` before calling `.Match(substring)` on the regexes, based on how those multi-byte Unicode encodings encode ASCII-range characters.

Aside from that, as noted, the current code also doesn't honor the proposed precedence rules.  (I am aware that its fundamentals predate your modifications, @CarloToso.) 

If we had explicit BOM detection, we could then simply skip the regex matching if a BOM is present.

iSazonov replied:
@mklement0 Thanks! So simplest fix is TryConvertToXml should use raw stream instead of string so that we delegate decoding to XmlReader, right?

As for precedence rules, if we agree to decode XML from raw stream this will resolve this too.

mklement0 replied:
Good point; delegating to `System.Xml.XmlDocument.Load(System.IO.Stream inStream)` with a raw byte stream is an option, but the above is still necessary for an HTML `<meta>` element.

However, this would result in _difference_ precedence rules, because the file/raw-bytes `.Load()` overload actually gives precedence to the `encoding` attribute in an XML declaration over a BOM - even though that can _break_ the parsing (e.g., if a single-byte encoding specified via `charset` conflicts with a multi-byte BOM).

In light of all this, perhaps it's better to change the precedence rules to prioritize explicit _out-of-band_ information, i.e. a `charset` spec. in the `Content-Type` header field, as follows, across all mediate types:

* Look for a (valid) `charset` attribute in the `Content-Type` header field, and, if present, honor it - no further investigation needed.

* Examine the first four raw bytes to see if they constitute a BOM - see https://en.wikipedia.org/wiki/Byte_order_mark#Byte_order_marks_by_encoding; restricting this to UTF-8 and the UTF-16 and UTF-32 variants is probably sufficient.

  * If so, always honor the specified encoding - and, in the case of XML, use _string_-based XML parsing (in which case the XML declaration's `encoding` attribute does NOT matter) - no further investigation needed.

* If not, look for an encoding specified via `<meta>` elements or XML-declarations and, if found, honor it (in the XML case, the raw-byte `.Load()` overload can then be used) - no further investigation needed.

* If there is none, default to UTF-8

CarloToso replied:
Unfortunately if we change the precedence rules we fail to fix https://github.com/PowerShell/PowerShell/issues/11547, charset=utf8, actual encoding=utf8withBom. I think we should check for the BOM and then for the charset

mklement0 replied:
@CarloToso, I don't think we need to handle this case, as it amounts to an _over-specification_ that we shouldn't be expect to compensate for:

If `charset=utf8` is present in the `Content-Type` header field _and_ there is (inappropriately) _also_ a UTF-8 BOM, I think it's fine (and appropriate) to parse this BOM as the actual character it represents, i.e. as [`U+FEFF`](http://www.fileformat.info/info/unicode/char/feff)) (ZERO WIDTH NO-BREAK SPACE)

iSazonov replied:
Our code resembles scraps. It should be made cleaner by avoiding breaking changes.
So first we call TryProcessFeedStream that is XML parsing of RSS/Feed based on raw stream. Then on fail we try the same thing again but trying to guess the response type and the encoding. It is not clear why we ignore the response type when we try to parse RSS/Feed. I believe we could remove one extra step.

mklement0 replied:
@iSazonov, good point about breaking changes.
Note that I am not deeply immersed in this, so do tell me if I'm getting something wrong:

Looking at the old code (https://github.com/PowerShell/PowerShell/blob/420c29d12c65e49a8d76909162286c24eee11d17/src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/StreamHelper.cs#L389), I see that the _out-of-band_ information - i.e. a `charset` attribute - currently already takes precedence, short-circuiting further investigation.

So we need to stick with that to avoid a breaking change, but do note that it contradicts the relevant RFC, as noted in https://github.com/PowerShell/PowerShell/issues/11547#issuecomment-637034379
Rethinking my over-specification statement: perhaps the right thing to do, if _both_ out-of-band encoding information and a BOM are present, is to remove the BOM _if the indicated / implied encoding is the same_.

In the absence of out-of-band information, the question is then how the handle the _in-band_ information - BOM and/or `encoding` / `charset` attribute in `<?xml>` declaration / `<meta>` HTML element.

The current code searches _only_ for the latter, and does so _incorrectly_, as discussed (no support for 2+-byte Unicode encodings). 

To fix this:

* For _XML_-based in-band encoding information, we can delegate to the .NET APIs based on raw byte streams.
* For the HTML `<meta>` case, we still need to dour own text-based matching, but it requires something like the `.Replace("\0")` approach mentioned above.

This too avoids a breaking change, though (like now) it requires the relatively expensive parsing of the first 1K / letting a .NET method call potentially fail (for potential XML content).

Giving precedence to manual BOM detection would perform better at least with non-XML content, but the only way  this would _not_ amount to a breaking change is _if the presence of a BOM currently breaks things_, which is the case, right?

Manual BOM detection is still necessary, at least in the absence of out-of-band encoding information and possibly also XML-declaration / HTML `<meta>` in-band information.

iSazonov replied:
.Net does encoding detection for XML https://source.dot.net/#System.Private.Xml/System/Xml/Core/XmlTextReaderImpl.cs,2895
So we have no need to do the same but worse.
Since first we try Feeds/Atom as XML we can simplify the code and first try XML reading then fallback to Json.

CarloToso replied:
@iSazonov we only try Feeds/Atom as XML in `Invoke-RestMethod` not in `Invoke-WebRequest`

iSazonov replied:
This does not prevent my suggestion.

CarloToso commented:
@iSazonov It seems I was able to make it work correctly

CarloToso commented:
@iSazonov Can we merge this or do I have to change something?

CarloToso commented:
@stevenebutler could you help me resolve this merge conflict?

stevenebutler commented:
> @stevenebutler could you help me resolve this merge conflict?

I can try, but I'm not sure how I would do that in github. Is it conflicting with the changes I made for ctrl-c, or failing the tests written for it?

stevenebutler committed:
Pr/19379 (#1)

* Remove FormObject.cs and FormObjectCollection.cs (#19383)

* Exclude redundant parameter aliases from completion results (#19382)

* Revert "Remove FormObject.cs and FormObjectCollection.cs (#19383)" (#19387)

This reverts commit 190c99a416b6f3c7e48066eadf3943d4ddccb2d8.

* Add the parameter `-RelativeBasePath` to `Resolve-Path` (#19358)

* Fix a crash in the type inference code (#19400)

* Remove GetResponseObject (#19380)

* Add `-Environment` parameter to `Start-Process` (#19374)

* Add `-Environment` parameter to `Start-Process`

* address codefactor

* fix test for Windows

* handle case where value is $null to remove env var

* change variables to make it more clear what the test is doing

* Add PoolNames variable group to compliance pipeline (#19408)

* Improve package management acceptance tests by not going to the gallery (#19412)

* Skip VT100 tests on Windows Server 2012R2 as console does not support it (#19413)

* Update the `ICommandPredictor` interface to reduce boilerplate code from predictor implementation (#19414)

* Enable type conversion of `AutomationNull` to `$null` for assignment (#19415)

* Remove code related to `#requires -pssnapin` (#19320)

* Support CTRL-C when reading data and connection hangs for `Invoke-RestMethod` and `Invoke-WebRequest` (#19330)

* Update to the latest NOTICES file (#19332)

* Update the cgmanifest (#19459)

* WIP: Harden default command test. (#19416)

CarloToso committed:
Merge branch 'master' into encodinghelper

CarloToso committed:
Merge branch 'master' into encodinghelper

CarloToso committed:
fix nullable

doctordns commented:
What is the status of this PR??

CarloToso commented:
@doctordns The code is complete and awaiting further review

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/StreamHelper.cs:
```diff
@@ -337,67 +337,23 @@ internal static void SaveStreamToFile(Stream stream, string filePath, PSCmdlet c
         {
-            StringBuilder result = new(capacity: ChunkSize);
-            Decoder decoder = encoding.GetDecoder();
+            bool isDefaultEncoding = !TryGetEncodingFromCharset(characterSet, out Encoding encoding);
```
The precedence of encoding detection is unclear to me.
By looking at the code, when `characterSet` is specified, `isDefaultEncoding` will be `false`, and
- meta element will not be checked
- encoding detected from BOM will take precedence, then the encoding resolved from `characterSet`

when `characterSet` is not specified, `isDefaultEncoding` will be `true`, and
- meta element will be checked
- character set from the meta element will take precedence, then the encoding detected from BOM, and then the default encoding

As is shown, the encoding detection is inconsistent -- if the `characterSet` specified in header is lower precedence than BOM, then why the `characterSet` specified in meta element is higher than BOM?

Can you please summarize the precedence you use and the reason for that (_e.g. to avoid breaking change? RFC defined? If they are contradicted with each other, then maybe we should make a breaking change to adhere to RFC?_) The summarization needs to be put in the code as comment.

daxian-dbw replied:
> Can you please summarize the precedence you use and the reason for that (e.g. to avoid breaking change? RFC defined? If they are contradicted with each other, then maybe we should make a breaking change to adhere to RFC?) The summarization needs to be put in the code as comment.

@CarloToso I see that you've made changes to the encoding detection. But can you please summarize the precedence and the reason for using that precedence?

CarloToso committed:
Add comments to code; better name for variable; encoding precedence to BOM

CarloToso committed:
Merge branch 'master' into encodinghelper

daxian-dbw commented:
@CarloToso The CIs are failing due to a compilation error. Can you please fix it?

CarloToso commented:
> @CarloToso The CIs are failing due to a compilation error. Can you please fix it?

There were some extensive changes in #19558, @stevenebutler could you help me once more?

stevenebutler commented:
> > @CarloToso The CIs are failing due to a compilation error. Can you please fix it?
> 
> There were some extensive changes in #19558, @stevenebutler could you help me once more?

Hi @CarloToso - If it's still an issue I may be able to have a look once I'm on vacation in a week or two.

stevenebutler committed:
Make encoding changes handle network stall timeouts

Added extension methods to StreamReader that will add a timeout
to each stream read if a timeout property is set in IWR/IRM.

stevenebutler commented:
> > > @CarloToso The CIs are failing due to a compilation error. Can you please fix it?
> > 
> > 
> > There were some extensive changes in #19558, @stevenebutler could you help me once more?
> 
> Hi @CarloToso - If it's still an issue I may be able to have a look once I'm on vacation in a week or two.

Hi @CarloToso  - I have made a PR on your fork with fixes for this PR

CarloToso committed:
Merge pull request #2 from stevenebutler/encodinghelper-patch

Make encoding changes handle network stall timeouts

CarloToso committed:
Merge branch 'PowerShell:master' into encodinghelper

stevenebutler committed:
Add ConfigureAwait(false) to awaits (#3)

This is needed to stop web methods from deadlocking when windows forms are loaded from
within the PowerShell process.

Jaykul commented:
This fix has been stuck for over a year, because ... what?
